### PR TITLE
basic docs for nested column query functions

### PR DESCRIPTION
**Pull request by clintropolis**
### Description
First round of docs for #12753, including native expression table, SQL functions table for web-console auto-complete, and virtual column documentation (written assuming #12920 goes in, may need adjustment pending review of that PR). Still missing a lot of stuff, so more to come in follow-up PRs.

Screen Shot 2022-08-18 at 2 08 34 PM

Also adds documentation for missing 'list filtered' virtual column.

### CODE REVIEW

File: docs/querying/sql-json-functions.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  The format of the tables that describe the functions and operators
+  should not be changed without updating the script create-sql-docs
+  in web-console/script/create-sql-docs, because the script detects
+  patterns in this markdown file and parse it to TypeScript file for web console
+-->
+
+Druid supports nested columns, which provide optimized storage and indexes for nested data structures. These JSON
+functions provide facilities to extract, transform, and create `COMPLEX<json>` values.
+
+| function | notes |
```
Comment (vtlim) [inline docs/querying/sql-json-functions.md:36]:
```suggestion
| Function | Notes |
```
Capitalization throughout the docs is not standardized but default towards sentence case in table headings

File: docs/misc/math-expr.md [hunk 2]
```diff
@@ -227,6 +226,34 @@ map((x) -> x + 1, x)
...
+| json_object(expr1, expr2[, expr3, expr4 ...]) | Construct a `COMPLEX<json>` with alternating 'key' and 'value' arguments|
+| parse_json(expr) | Deserialize a JSON `STRING` into a `COMPLEX<json>` to be used with expressions which operate on `COMPLEX<json>` inputs. Non-`STRING` input or invalid JSON will result in an error. |
+| try_parse_json(expr) | Deserialize a JSON `STRING` into a `COMPLEX<json>` to be used with expressions which operate on `COMPLEX<json>` inputs. Non-`STRING` input or invalid JSON will result in a `NULL` value. |
+| to_json_string(expr) | Convert a `COMPLEX<json>` input into a JSON `STRING` value |
+| json_keys(expr, path) | get array of field names in `expr` at the specified JSONPath `path`, or null if the data does not exist or have any fields |
+| json_paths(expr) | get array of all JSONPath paths available in `expr` |
+
+### JSONPath syntax
+
+Druid supports a small, simplified subset of the [JSONPath syntax](https://github.com/json-path/JsonPath/blob/master/README.md) operators, primarily limited to extracting individual values from nested data structures.
```
Comment (vtlim) [inline docs/misc/math-expr.md:246]:
```suggestion
Druid supports a subset of the [JSONPath syntax](https://github.com/json-path/JsonPath/blob/master/README.md) operators, primarily limited to extracting individual values from nested data structures.
```

File: docs/querying/sql-json-functions.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+|`PARSE_JSON(expr)`|Parses a string type `expr` into a `COMPLEX<json>` object. This operator deserializes JSON values when processing them, translating stringified JSON into a nested structure. Non-`STRING` input or invalid JSON will result in an error.|
+|`TRY_PARSE_JSON(expr)`|Parses a string type `expr` into a `COMPLEX<json>` object. This operator deserializes JSON values when processing them, translating stringified JSON into a nested structure. Non-`STRING` input or invalid JSON will result in a `NULL` value.|
+|`TO_JSON_STRING(expr)`|Casts an `expr` of any type into a `COMPLEX<json>` object, then serializes the value into a JSON string.|
+
+### JSONPath syntax
+
+Druid supports a small, simplified subset of the [JSONPath syntax](https://github.com/json-path/JsonPath/blob/master/README.md) operators, primarily limited to extracting individual values from nested data structures.
```
Comment (vtlim) [inline docs/querying/sql-json-functions.md:49]:
```suggestion
Druid supports a subset of the [JSONPath syntax](https://github.com/json-path/JsonPath/blob/master/README.md) operators, primarily limited to extracting individual values from nested data structures.
```

File: docs/querying/sql-json-functions.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+|`['<name>']`| Child element in bracket notation. |
+|`[<number>]`| Array index. |
+
+Consider the following example input JSON:
+
+```json
+{"x":1, "y":[1, 2, 3]}
+```
+
+- To return the JSON object:<br>
```
Comment (vtlim) [inline docs/querying/sql-json-functions.md:64]:
```suggestion
- To return the entire JSON object:<br>
```

File: docs/querying/sql-json-functions.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  `$`      -> `{"x":1, "y":[1, 2, 3]}`
+- To return the value of a key "x":<br>
```
Comment (vtlim) [inline docs/querying/sql-json-functions.md:66]:
```suggestion
- To return the value of the key "x":<br>
```

Comment (vtlim) [reply docs/querying/sql-json-functions.md:66]: nice examples!

File: docs/querying/virtual-columns.md [hunk 6]
```diff
@@ -80,4 +82,106 @@ The expression virtual column has the following syntax:
 |--------|-----------|---------|
 |name|The name of the virtual column.|yes|
```
Comment (vtlim) [inline docs/querying/virtual-columns.md:83]: This table should include the `type` property as required. Description can be something like "Type of the virtual column. Set to `expression` for an expression virtual column."

Comment (vtlim) [reply docs/querying/virtual-columns.md:83]: Ditto for the `nested-field` virtual column and `mv-filtered` virtual column

File: docs/querying/virtual-columns.md [hunk 7]
```diff
@@ -80,4 +82,106 @@ The expression virtual column has the following syntax:
...
-|outputType|The expression's output will be coerced to this type. Can be LONG, FLOAT, DOUBLE, or STRING.|no, default is FLOAT|
+|outputType|The expression's output will be coerced to this type. Can be LONG, FLOAT, DOUBLE, STRING, ARRAY types, or COMPLEX types.|no, default is FLOAT|
+
+
+### Nested field virtual column
+
+The nested field virtual column is an optimized virtual column that can provide direct access into various paths of
+a `COMPLEX<json>` column, including using their indexes.
+
+Syntax (all 3 of these virtual columns produce the same output):
```
Comment (vtlim) [inline docs/querying/virtual-columns.md:93]:
```suggestion
You can define a nested field virtual column with any of the following syntaxes.
The examples differ in how to access a specific value from the `COMPLEX<json>`.
All examples produce the same output value.
Reference a virtual column by its output name, such as `v0`, `v1`, or `v2`.
```

File: docs/querying/virtual-columns.md [hunk 8]
```diff
@@ -80,4 +82,106 @@ The expression virtual column has the following syntax:
...
+      ]
+    }
+```
+
+|property|description|required?|
+|--------|-----------|---------|
+|columnName|The name of the virtual column.|yes|
+|outputName|The name of the virtual column.|yes|
+|expectedType|The name of the virtual column.|yes|
+|pathParts|The name of the virtual column.|yes|
```
Comment (vtlim) [inline docs/querying/virtual-columns.md:142]: Update description?

File: docs/querying/virtual-columns.md [hunk 9]
```diff
@@ -80,4 +82,106 @@ The expression virtual column has the following syntax:
...
+|processFromRaw|If set to true, the virtual column will process the "raw" JSON data to extract values rather than using an optimized "literal" value selector. This option allows extracting non-literal values (such as nested JSON objects or arrays) as a `COMPLEX<json>` at the cost of much slower performance.|No, default false|
+|path|'JSONPath' or 'jq' syntax path. One of `path` or `pathParts` must be set|no, if `pathParts` is defined|
+|useJqSyntax||no, default is false|
```
Comment (vtlim) [inline docs/querying/virtual-columns.md:145]: Missing description?

File: docs/querying/virtual-columns.md [hunk 10]
```diff
@@ -80,4 +82,106 @@ The expression virtual column has the following syntax:
...
+
+#### Nested path part
+|property|description|required?|
+|--------|-----------|---------|
+|type|Must be 'field' or 'arrayElement'|yes|
```
Comment (vtlim) [inline docs/querying/virtual-columns.md:150]:
```suggestion
|type|Must be `field` or `arrayElement`. Use `field` when accessing a specific field in a nested structure. Use `arrayElement` when accessing a specific integer position of an array.|yes|
```
Array position starts with 0?

File: docs/querying/virtual-columns.md [hunk 11]
```diff
@@ -80,4 +82,106 @@ The expression virtual column has the following syntax:
...
+|field|The name of the 'field' in a 'field' `type` path part|yes, if `type` is 'field'|
+|index|The array element index if `type` is `arrayElement`|yes, if `type` is 'arrayElement'|
+
+This virtual column is used for the SQL operators `JSON_VALUE` (if `processFromRaw` is set to false) or `JSON_QUERY`
+(if it is true), and accepts 'JSONPath' or 'jq' syntax string representations of paths, or a parsed
```
Comment (vtlim) [inline docs/querying/virtual-columns.md:155]:
```suggestion
(if `processFromRaw` is true), and accepts 'JSONPath' or 'jq' syntax string representations of paths, or a parsed
```

File: web-console/script/create-sql-docs.js [hunk 12]
```diff
@@ -63,6 +63,7 @@ const readDoc = async () => {
     await fs.readFile('../docs/querying/sql-scalar.md', 'utf-8'),
     await fs.readFile('../docs/querying/sql-aggregations.md', 'utf-8'),
     await fs.readFile('../docs/querying/sql-multivalue-string-functions.md', 'utf-8'),
+    await fs.readFile('../docs/querying/sql-json-functions.md', 'utf-8'),
```
Comment (vtlim) [inline web-console/script/create-sql-docs.js:66]: Not sure if `MINIMUM_EXPECTED_NUMBER_OF_FUNCTIONS` should be updated as well? (L26)

File: docs/querying/virtual-columns.md [hunk 13]
```diff
@@ -80,4 +82,106 @@ The expression virtual column has the following syntax:
...
+|--------|-----------|---------|
+|columnName|The name of the virtual column.|yes|
+|outputName|The name of the virtual column.|yes|
+|expectedType|The name of the virtual column.|yes|
+|pathParts|The name of the virtual column.|yes|
+|processFromRaw|If set to true, the virtual column will process the "raw" JSON data to extract values rather than using an optimized "literal" value selector. This option allows extracting non-literal values (such as nested JSON objects or arrays) as a `COMPLEX<json>` at the cost of much slower performance.|No, default false|
+|path|'JSONPath' or 'jq' syntax path. One of `path` or `pathParts` must be set|no, if `pathParts` is defined|
+|useJqSyntax||no, default is false|
+
+#### Nested path part
```
Comment (vtlim) [inline docs/querying/virtual-columns.md:147]:
```suggestion
#### Nested path part
Specify `pathParts` as an array of objects that describe each component of the path to traverse. Each object can take the following properties:
```

File: docs/querying/virtual-columns.md [hunk 14]
```diff
@@ -80,4 +82,106 @@ The expression virtual column has the following syntax:
...
+|property|description|required?|
+|--------|-----------|---------|
+|type|Must be 'field' or 'arrayElement'|yes|
+|field|The name of the 'field' in a 'field' `type` path part|yes, if `type` is 'field'|
+|index|The array element index if `type` is `arrayElement`|yes, if `type` is 'arrayElement'|
+
+This virtual column is used for the SQL operators `JSON_VALUE` (if `processFromRaw` is set to false) or `JSON_QUERY`
```
Comment (vtlim) [inline docs/querying/virtual-columns.md:154]: This paragraph might be more useful at the beginning of the section (before the syntax code blocks)

File: docs/querying/virtual-columns.md [hunk 15]
```diff
@@ -80,4 +82,106 @@ The expression virtual column has the following syntax:
...
+(if it is true), and accepts 'JSONPath' or 'jq' syntax string representations of paths, or a parsed
+list of "path parts" in order to determine what should be selected from the column.
+
+Type information for nested fields is absent at higher levels (it is contained within the segment, but not to segment
+metadata queries or the SQL planner), so `expectedType` provides the context for how something is being used, e.g. an
+aggregators default type or an explicit cast, or, if using the 'RETURNING' syntax which explicitly specifies type.
```
Comment (vtlim) [inline docs/querying/virtual-columns.md:160]:
```suggestion
aggregators default type or an explicit cast, or, if using the 'RETURNING' syntax in `JSON_VALUE` to explicitly specify type.
```

File: docs/querying/sql-data-types.md [hunk 16]
```diff
@@ -33,7 +33,7 @@ Columns in Druid are associated with a specific data type. This topic describes
 
 Druid natively supports five basic column types: "long" (64 bit signed int), "float" (32 bit float), "double" (64 bit
 float) "string" (UTF-8 encoded strings and string arrays), and "complex" (catch-all for more exotic data types like
-hyperUnique and approxHistogram columns).
+json, hyperUnique, and approxHistogram columns).
```
[the remark this answers is not in the document]
Comment (clintropolis) [reply docs/querying/sql-data-types.md:36]: not really sure, i felt compelled to add json because i talk about it later on

File: docs/querying/virtual-columns.md (hunk 8 above)
Comment (clintropolis) [reply docs/querying/virtual-columns.md:142]: heh oops

File: docs/querying/virtual-columns.md (hunk 15 above)
Comment (clintropolis) [reply docs/querying/virtual-columns.md]: i removed this section and just expanded the information for `expectedType` to indicate that druid coerces values to the specified type if it is supplied.

File: docs/querying/virtual-columns.md (hunk 14 above)
Comment (clintropolis) [reply docs/querying/virtual-columns.md]: moved, though it's not really necessary to care how SQL planner uses this native virtual column to implement its functions, so it could also be removed probably. Its just adding flavor about what this thing is for I guess.

File: docs/querying/virtual-columns.md (hunk 13 above)
Comment (clintropolis) [reply docs/querying/virtual-columns.md:153]: added :+1:

File: docs/querying/sql-functions.md [hunk 17]
```diff
@@ -647,6 +647,46 @@ Parses `address` into an IPv4 address stored as an integer.
...
+
+Returns an array of all paths which refer to literal values in a `COMPLEX<json>` typed `expr`, in JSONPath format.
+
+## JSON_QUERY
+
+**Function type:** [JSON](sql-json-functions.md)
+
+`JSON_QUERY(expr, path)`
+
+Extracts a `COMPLEX<json>` value from a `COMPLEX<json>` typed `expr`, at the specified `path`.
```
[the remark this answers is not in the document]
Comment (clintropolis) [reply docs/querying/sql-functions.md]: i replace all fragments of the form 
> from a `COMPLEX<json>` typed `expr`
with
> `from `expr`
since the former was no longer actually even accurate

File: docs/querying/sql-data-types.md [hunk 18]
```diff
@@ -112,3 +112,17 @@ When `druid.expressions.useStrictBooleans = false` (the default mode), Druid use
...
+interacted with using [JSON functions](sql-json-functions.md), which can extract nested values, parse from string,
+serialize to string, and to create new `COMPLEX<json>` structures.
+
+`COMPLEX` types in general currently have limited functionality outside of the use of the specialized functions which
+understand them, and so have undefined behavior when:
+* grouping on complex values
+* filtered directly on complex values, e.g. `WHERE json is NULL`
+* used as inputs to aggregators without specialized handling for a specific complex type
+
+In many cases, these functions are provided for translating these `COMPLEX` value types a `STRING`, which serves as
```
Comment (vtlim) [inline docs/querying/sql-data-types.md:127]:
```suggestion
In many cases, functions are provided to translate `COMPLEX` value types to `STRING`, which serves as a
```

File: docs/querying/sql-functions.md [hunk 19]
```diff
@@ -647,6 +647,46 @@ Parses `address` into an IPv4 address stored as an integer.
...
+
+Returns an array of field names from `expr` at the specified `path`.
+
+## JSON_OBJECT
+
+**Function type:** [JSON](sql-json-functions.md)
+
+`JSON_OBJECT(KEY expr1 VALUE expr2[, KEY expr3 VALUE expr4, ...])`
+
+Constructs a new `COMPLEX<json>` object. The `KEY` expressions must evaluate to string types. The `VALUE` expressions can be composed of any input type, including other `COMPLEX<json>` values. `JSON_OBJECT` can accept alternating key-value pairs separated by colons. The following syntax is equivalent: `JSON_OBJECT(expr1:expr2[, expr3:expr4, ...])`.
```
Comment (vtlim) [inline docs/querying/sql-functions.md:664]:
```suggestion
Constructs a new `COMPLEX<json>` object. The `KEY` expressions must evaluate to string types. The `VALUE` expressions can be composed of any input type, including other `COMPLEX<json>` values. `JSON_OBJECT` can accept colon-separated key-value pairs. The following syntax is equivalent: `JSON_OBJECT(expr1:expr2[, expr3:expr4, ...])`.
```
Technically the key-value pairs are separated by commas but it might be too wordy to say "colon-separated" then "separated by commas"

File: docs/querying/sql-data-types.md [hunk 20]
```diff
@@ -112,3 +112,17 @@ When `druid.expressions.useStrictBooleans = false` (the default mode), Druid use
 When `druid.expressions.useStrictBooleans = true`, Druid uses three-valued logic for
 [expressions](../misc/math-expr.md) evaluation, such as `expression` virtual columns or `expression` filters.
 However, even in this mode, Druid uses two-valued logic for filter types other than `expression`.
+
+## Nested columns
+Druid supports storing nested data structures in segments using the native `COMPLEX<json>` type. This data can be
+interacted with using [JSON functions](sql-json-functions.md), which can extract nested values, parse from string,
```
Comment (vtlim) [inline docs/querying/sql-data-types.md:118]:
```suggestion
Druid supports storing nested data structures in segments using the native `COMPLEX<json>` type. You can interact
with this data using [JSON functions](sql-json-functions.md), which can extract nested values, parse from string,
```

File: docs/querying/virtual-columns.md (hunk 14 above)
Comment (vtlim) [reply docs/querying/virtual-columns.md]: ah true. the helpful part for the intro was more so on how you can use JSONPath or jq syntax to extract part of the json

File: docs/misc/math-expr.md [hunk 21]
```diff
@@ -227,6 +226,34 @@ map((x) -> x + 1, x)
...
+| function | description |
+|---|---|
+| json_value(expr, path) | Extract a Druid literal (`STRING`, `LONG`, `DOUBLE`) value from `expr` using JSONPath syntax of `path` |
+| json_query(expr, path) | Extract a `COMPLEX<json>` value from `expr` using JSONPath syntax of `path` |
+| json_object(expr1, expr2[, expr3, expr4 ...]) | Construct a `COMPLEX<json>` with alternating 'key' and 'value' arguments|
+| parse_json(expr) | Deserialize a JSON `STRING` into a `COMPLEX<json>`. If the input is not a `STRING` or it is invalid JSON, this function will result in an error.|
+| try_parse_json(expr) | Deserialize a JSON `STRING` into a `COMPLEX<json>`. If the input is not a `STRING` or it is invalid JSON, this function will result in a `NULL` value. |
+| to_json_string(expr) | Convert `expr` into a JSON `STRING` value |
+| json_keys(expr, path) | get array of field names from `expr` at the specified JSONPath `path`, or null if the data does not exist or have any fields |
+| json_paths(expr) | get array of all JSONPath paths available from `expr` |
```
Comment (vtlim) [inline docs/misc/math-expr.md:242]:
```suggestion
| json_keys(expr, path) | Get array of field names from `expr` at the specified JSONPath `path`, or null if the data does not exist or have any fields |
| json_paths(expr) | Get array of all JSONPath paths available from `expr` |
```

**Review by vtlim — APPROVED** [review]
LGTM 🦖
